design-proposals/cluster-orch-integrating-k8s-into-emt.md

```diff
@@ -0,0 +1,246 @@
+
+The sequence diagram below illustrates the API call flow for host registration when `Cluster Join Automatically` is enabled.
+
+```mermaid
+sequenceDiagram
+    actor User
+    User->>+UI: Register a new host
+    opt
+    UI->>+Cluster Manager: Create a cluster
+    Cluster Manager-->>-UI: Success
+    end
+    UI->>+Infra Manager: Register host
+    Infra Manager-->>-UI: Success
+    UI->>+Infra Manager: Create instance
+    Infra Manager-->>-UI: Success
+    UI->>+Cluster Manager: Add host to the cluster
+    Cluster Manager-->>-UI: Success
+    UI-->>-User: Success
+```
+
+*[todo: discuss transaction management and failure handling]*
+*[todo: update sequence diagram with failure scenario]*
+
+A new cluster is created only when the user does not provide an existing cluster name and opts for the default action, which is creating a new single-node cluster. In the final step, the host is added to either the user-provided existing cluster or the newly created cluster.
+
+*[todo: add another sequence diagram for host connect, cluster reconciliation pause/unpause]*
+
+#### Manual Cluster Creation
+
+Manual cluster creation is an existing method where users create a cluster by selecting a template and hosts. Before we talk about manual cluster creation workflow, it is important to understand the coupling between EMT image version and cluster template introduced by integrating Kubernetes into EMT. For instance, creating a K3s cluster with a K3s v1.32 template may not work as expected on a EMT machine provisioned with K3s v1.30. Such discrepancies can occur when Edge Orchestration undergoes multiple upgrades, resulting in a mix of multiple versions of cluster templates and EMT machines. While this issue is relevant for K3s on EMT only, given that this combination is the primary use case, it is important to address it properly in our user workflow. There are a few options to mitigate this issue.
+
+**Option 1)** Restrict cluster creation to onboarded hosts only, excluding provisioned hosts. The Cluster Manager automatically selects the appropriate EMT version and requests instance creation from the Infra Manager to install the OS and Kubernetes.
+
+**Option 2)** Implement logic and a new API within the Cluster Manager to check version compatibility and return eligible hosts for the selected cluster template.
+
+**Option 3)** Implement logic within the Cluster Manager to check version compatibility and initiate a request to the Infra Manager for an A/B update to a compatible EMT version before bootstrapping Kubernetes. (A/B update to lower version is allowed?)
```

Comment:
Does this count as a separate option or an implementation detail. I guess the point is that we need to have a version on K8s version (either active/passive partition) to be able to cluster on EMT. If a version is available the suggestion here is to chose that partition for bootstrapping the node.

Reply:
yeah to me this seems to be an additional feature to Option1

Reply:
Look "Delete Cluster" to see difference.